### ISSUE TITLE
Auto open completion menu; Press 'Enter' only once on the completion menu item

### ISSUE THREAD

**Issue by spcmd** [mason-larobina/luakit#297]
Is there a way to open the completion menu automatically (like Vimperator)? 

Also, I'd like to select a menu item by pressing the Enter key **once**, so when I hit Enter on a completion menu item I'd like to go to the url instead of just writing the url to the command bar and wait for another 'Enter'.

**Comment by spcmd** [comment]
Meanwhile I figured out the 'Enter' key stuff:

completion.lua (at line 158):
```
activate = function (w, text)
        local pos = w.ibar.input.position
        if string.sub(text, pos+1, pos+1) == " " then pos = pos+1 end
        w:enter_cmd(text, { pos = pos })
        -- activate the selected item instead of waiting for another 'Enter'
        w:activate()
    end,
```

I think it should be in the wiki. Or am I missing something?

**Comment by xkpd3** [comment]
I actually think this should be the standard behavior. What do you think @aidanholm?

**Comment by aidanholm** [comment]
Completion is currently a little wonky, and could definitely be improved. I did some work a while back on getting multi-mode support (kinda like Emacs' major and minor modes) working in a personal branch, which if iirc made a lot of things like completion simpler. I'll have to brush that off after the next release.

I think that opening the menu by default and only needing a single key press are both good defaults.

**Comment by xkpd3** [comment]
I just saw, that I was not accurate. I was referring only to the "enter-once" part. I don't like the menu to open automatically. Menu on tab I think is good. And it should also only open when more than one match is found.

The main reason (for me) is, that I can use the browser with people around me without accidentally revealing parts of my browsing history :)

**Comment by aidanholm** [comment]
I think we could add some configuration options to this one:
```lua
command.auto_open_completion_menu = true -- or false
completion.autocomplete_on_single_match = true -- or false
```

**Comment by aidanholm** [comment]
Enter-once has been added with the merge of the new completion engine at f8c8e76b

**Comment by aidanholm** [comment]
From @schvabodka-man:
> Hello there again. That would be really nice if there would be "real-time" autocompletition for commands and history as in qutebrowser(for example). That wouldn't be hard to implement. At the moment you already have autocompletition code. It's "fired" when user are pressing tab key. You just need to implement some listener that would "fire" autocompletition every time input in luakit "command line" is changed. This would really improve usability, because clicking tab key is little bit annoying

**Comment by aidanholm** [comment]
I think the best way to add the remaining features would probably be to just merge completion functionality into the command mode proper, as maintaining the separation is too hacky.

**Comment by arcnmx** [comment]
I'd like to add to this that the completion engine doesn't expose any way to extend it (`completers` is a local table, and there's no `add_completer()` sort of API). Each completer for the various modules is custom-implemented inside of `completion.lua`, which seems less than ideal - and also pulls in modules that the user may otherwise not have enabled (like history tracking, for example).

Not sure if this should be a separate issue but it seems like something that would be resolved on its own once these proposed changes are introduced.
